includes/rm_and_kill.py

```
@@ -0,0 +1,253 @@
+        if missing_ok:
+            return None
+        else:
+            # In order to get correct error
+            os.kill(pid, signal.SIGTERM)
+
+    # Otherwise try `retries` times
+    i = 0
+    while psutil.pid_exists(pid):
+        # Try without guard in order to get correct error
+        if i == retries:
+            os.kill(pid, signal.SIGTERM)
+            break
+
+        try:
+            os.kill(pid, signal.SIGTERM)
+        except OSError as e:
+            pass
+
+        i += 1
+
+@contextmanager
+def handler():
+    # Put here what would ordinarily go in the `__enter__` method
+    # In this case, there's nothing to do
+    try:
+        yield None# You can return something if you want, that gets picked up in the 'as'
+    except Exception as e:
+        print("no not possible")
+    finally:
+        print("Ok I caught you")
+
+
+@contextmanager
+def create_locked_file(fpath: Path = None) -> Tuple[str, int]:
+    """
+
+    >>> with create_locked_file() as (fpath, pid):
+    ...     os.remove(fpath)  # doctest: +ELLIPSIS
+    Traceback (most recent call last):
+     ...
+    PermissionError: [WinError 32] ...
+
+    """
+
+    if fpath is not None:
+        if fpath.exists():
+            raise FileExistsError("Cannot create file, file already exists.")
+        fpath = Path(fpath).absolute()
+    else:
+        fpath = Path(tempfile.mktemp())
+
+    lockcode = textwrap.dedent(f'''
+        import sys
+        import time
+        f = open(r'{fpath}', 'w')
+        while True:
+            time.sleep(60)
+    ''')
+
+    try:
+        p = subprocess.Popen([sys.executable, "-c", lockcode])
+        while not fpath.exists():
+            pass
+
+        yield str(fpath.absolute()).lower(), p.pid
+
+    finally:
+        p.kill()
+        if not fpath.exists():
+            return
+
+        while True:
+            try:
+                os.remove(fpath)
+                break
+            except PermissionError:
+                pass
```

This will cause an busy loop (high CPU usage) when the file's permissions have changed for some reason. Not good.

---

Okay, I only use this in tests, but that is no excuse for a sloopy function. Do you think I should add a time.sleep or what is the best way to overcome a busy loop?

---

If it's only used in tests, but it in a separate file and write in the doc string "this is only used in tests".

Never mind busy loop, why do you need a loop at all? Just let the test fail if you fail to remove the file. Or if you really need to wait for the process to end, use p.kill() followed by p.join() or something like that. Worst case, if you really need to loop, add time.sleep() and a maximum number of retries before throwing an error.

---

Okay, I added that this is used for testing.